### PR TITLE
[WIP] Preference model

### LINKED ISSUE #996: [v0.2-Athena]: Post-train SLM for human preference signal extraction
## Description
Post-train a Small Language Model (SLM) specifically for extracting human preference signals.

## Context
Part of v0.2 Athena milestone - MoM Enhancements

## Reference
https://blog.vllm.ai/2026/01/05/vllm-sr-iris.html

### PR DESCRIPTION
**Pull request by ppppqp**
FIX https://github.com/vllm-project/semantic-router/issues/996

# Background
Since different models can have different performances on different tasks, there were some [researches](https://arxiv.org/pdf/2506.16655)  about creating an LLM router based on human preference. In v-LLM SR, we also listed preference signal as one of the signals we consider when making the routing decision. 
This PR implements the pipeline for such model, including training date cleaning, fine-tuning and evaluation on multiple benchmarks, effectively laying the foundation for future iterations.

# Design

## Training data preparation

In the Arch-Router paper, the authors used LLM to synthesize training data. They first curated a topic pool based on industry report, MMLU and API documentations. They used LLM to extend the topic pool to follow the Domain–Action taxonomy, and used LLM to synthesize conversations from each extended topic.

The approach in the PR is slightly different: we used the conversation data from [ShareGPT](https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/blob/main/ShareGPT_V3_unfiltered_cleaned_split.json), which is a common choice for LLM benchmark, as the base corpus. Then we try to label each conversation with a routing policy. The benefit for doing this is (or at least trying this) that ShareGPT's data is real human-AI interaction data. This is the data that is the closest to how model will be used in practice. This makes the model to learn the conversation distribution more effectively, and makes the evaluation benchmark more meaningful.

The methodology for generating the label for each conversation follows the open-set classification problem. 
For each conversation, we pipe the conversation to LLM and ask it to generate a policy that is most suitable for the conversation context, and then generate a list of negative policies that are plausible but do not match the conversation. After we processed all conversations, we now have a truth label and a list of negative labels for each conversation.

The complete process of generating dataset take about 2 to 3 hours. I generated 25k samples on the first run (used for MVP). One conversation in ShareGPT is splitted into multiple chunks for not exceeding model length. For faster iteration speed and control the label space, I always label the first chunk, and assumes the other chunks falls into the same category, which based on my observation it is usually the case. If we hold this assumption, the dataset can be extended to 50k samples, which matches the training data size mentioned in Arch Router.

The generated dataset is here: https://huggingface.co/datasets/ppppqp/vLLM-SR-Preference-V1

## Supervised Fine-tuning

We used the classic text generation as fine-tuning approach for MVP. For each sample, generate a random label space from the label pool (truth label + negative labels) and plug the conversation into the prompt template and ask the model to output the best matching label. We masks all prompt and only compute loss based on the final predicted label. To increase the robustness of the model, the label pool is sampled (between 4 labels to 16 labels) and randomized in label ordering.

To account for the case where none of the routing policy matches the conversation, we exclude the truth label from the label pool for about 6% (1/label_pool_size, which is 16) of all training samples. For those training samples, the truth now becomes "none_of_the_above", which is the hardcoded catch-all token for this scenario.

## Validation
We generate another 250 samples with the same process as training samples, and run them through the base model (Qwen3-0.6B) and the finetuned model, using the same prompt template as the training process. We test by checking the accuracy by counting the number of samples that the output of the model matches exactly the truth policy label. We also tested the "all-negative" case where the truth label is not included in the label space (where the model should output "none_of_the_above".

Below is the result (accuracy)
||base model|finetuned model|
|---|---|---|
|normal case|0.149|0.972|
|all negative case|0.469|0.988|

CLINC150
||base model|finetuned model|
|---|---|---|
|normal case|0.156|0.260|
|all negative case|/|/|

# Future Improvements

There are a lot to improve on.

### Inference paradigm
Currently the fine-tuning is framed as a text generation task due to the openness of label space. The benefit is pretty obvious: it's simple with great generality, which makes it the perfect candidate for MVP. 
However, it might worth thinking if it's better to be framed into a classification problem, where we can do softmax to get the calibrated confidence score of each label.

### Larger dataset
Training with all samples in ShareGPT (100k).

### DPO alignment
Something to explore

---

- [ ] Make sure the code changes pass the [pre-commit](https://github.com/vllm-project/semantic-router/blob/main/CONTRIBUTING.md) checks.
- [ ] Sign-off your commit by using <code>-s</code> when doing <code>git commit</code>
- [ ] Try to classify PRs for easy understanding of the type of changes, such as `[Bugfix]`, `[Feat]`, and `[CI]`.

<details>

<summary><b> Detailed Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to semantic-router! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[CLI]</code> for changes to the command-line interface tools.</li>
    <li><code>[Dashboard]</code> for changes to the dashboard or web UI.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Feat]</code> for new features in the cluster (e.g., autoscaling, disaggregated prefill, etc.).</li>
    <li><code>[Router]</code> for changes to the <code>vllm_router</code> (e.g., routing algorithm, router observability, etc.).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>Pass all linter checks. Please use <code>pre-commit</code> to format your code. See <code>README.md</code> for installation.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient tests to ensure the change is stay correct and robust. This includes both unit tests and integration tests.</li>
</ul>

<h3>DCO and Signed-off-by</h3>
<p>When contributing changes to this project, you must agree to the <a href="https://github.com/vllm-project/vllm/blob/main/DCO">DCO</a>. Commits must include a <code>Signed-off-by:</code> header which certifies agreement with the terms of the DCO.</p>
<p>Using <code>-s</code> with <code>git commit</code> will automatically add this header.</p>

<h3>What to Expect for the Reviews</h3>

### CODE REVIEW

**Comment by Xunzhuo** [comment]
Closing this superseded WIP. #1228 and #1368 delivered the preference E2E and contrastive preference classifier now present on main. This older training pipeline has not moved since January; any remaining training work should be a focused follow-up against the current preference contract.
